bsp/src/org/jetbrains/bsp/project/test/BspTestRunProfileState.scala

```diff
@@ -0,0 +1,145 @@
+import com.intellij.openapi.module.ModuleManager
+import com.intellij.openapi.project.Project
+import jetbrains.buildServer.messages.serviceMessages._
+import org.jetbrains.bsp.data.BspMetadata
+import org.jetbrains.bsp.protocol.BspCommunicationComponent
+import org.jetbrains.bsp.protocol.BspNotifications.{BspNotification, TaskFinish, TaskStart}
+import org.jetbrains.bsp.protocol.session.BspSession.BspServer
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits._
+
+class BspTestRunProfileState(val project: Project, val rc: BspTestRunConfiguration, val ex: Executor) extends RunProfileState {
+
+  val gson = new Gson()
+
+  class MProcHandler extends ProcessHandler {
+    override def destroyProcessImpl(): Unit = ()
+
+    override def detachProcessImpl(): Unit = {}
+
+    override def detachIsDefault(): Boolean = false
+
+    override def getProcessInput: OutputStream = null
+
+    def shutdown(): Unit = {
+      super.notifyProcessTerminated(0)
+    }
+  }
+
+  private def targets(): List[URI] = {
+    ModuleManager.getInstance(project).getModules.toList
+      .flatMap(BspMetadata.get(project, _))
+      .flatMap(x => x.targetIds.asScala.toList)
+  }
+
+  private def testRequest(server: BspServer): CompletableFuture[TestResult] = {
+    val targetIds = targets().map(uri => new BuildTargetIdentifier(uri.toString))
+    val params = new TestParams(targetIds.toList.asJava)
+    params.setOriginId(UUID.randomUUID().toString)
+    server.buildTargetTest(params)
+  }
+
+
+  def printProc(str: MessageWithAttributes)(implicit proc: ProcessHandler): Unit = {
+    proc.notifyTextAvailable(str + System.lineSeparator(), ProcessOutputType.STDOUT)
+  }
+
+  class ServiceMsg(val name: String, val map: Map[String, String]) extends MessageWithAttributes(name, map.asJava) {
+    def this(n: String) = this(n, Map())
+  }
+
+  case class NestedBspMsg(nested: AnyRef, msg: AnyRef)
+
+  type HasNested = {
+    def getDataKind: String
+    def getData: Object
+  }
+
+  private def extractNestedMessage(bspMessage: AnyRef): AnyRef = bspMessage match {
+    case x: HasNested@unchecked =>
+      val kind = x.getDataKind
+      val rawNested = x.getData
+      rawNested match {
+        case d: JsonObject => kind match {
+          case TEST_FINISH => gson.fromJson(d, classOf[TestFinish])
+          case TEST_TASK => gson.fromJson(d, classOf[TestTask])
+          case TEST_REPORT => gson.fromJson(d, classOf[TestReport])
+          case TEST_START => gson.fromJson(d, classOf[TestStart])
+          case _ => rawNested
+        }
+        case _ => rawNested
+      }
+    case x => x
+  }
+
+  class BspTestSession {
+    val startTime: mutable.Map[String, Long] = mutable.Map()
+  }
+
+  // TODO handle standard output of test cases
+  private def onBspNotification(proc: ProcessHandler, session: BspTestSession)(n: BspNotification): Unit = {
```

Comment:
Is there a way to handle BspTestSession in a more functional manner ?

Reply:
what do you have in mind?

Reply:
Here I mutate the session object, which has a mutable map inside. I thought there should be a way to handle it with immutable map.

Reply:
The alternative is to run org.jetbrains.bsp.protocol.BspCommunication#run with aggregator: `NotificationAggregator[A]` instead of `NotificationCallback`. That allows you to create a new/modified object on every call. I'm not sure if it would have much of an advantage here.